DM-32979: initial checkin of hermesk daemon mods

srp3rd opened the pull request.

srp3rd committed:
Initial check-in

srp3rd committed:
Add build.yaml

srp3rd committed:
Add "container" to docker-compose file

srp3rd committed:
Change "container" to "container_name"

srp3rd committed:
Rework VERSION

srp3rd committed:
Remove version number for build

srp3rd committed:
Update versions

srp3rd committed:
Add logging.exception lines
change WARN to INFO

ktlim commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,38 @@
+    - name: Build ingest image
+      working-directory: docker
+      run: |
+        docker-compose -f "docker-compose.yml" --env-file versions.env build $HERMESK_NAME
```
It looks like there isn't that much in the `docker-compose.yml`.  You might be able to use https://github.com/lsst-sqre/build-and-push-to-ghcr if you go directly to the `Dockerfile`.

srp3rd replied:
I tried, but was unable to get it to work.  Plus the current version of build-and-push-to-ghcr doesn't support env (at least according to a message today on slack).  In an effort to get this out, I'm fixing the `docker-compose` to `docker compose` and staying with this for now.  I'll add that change to another ticket.

ktlim commented on docker/rucio-daemons-hermesk/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+RUN pip install retrying
+
+RUN mkdir -p /etc/grid-security/certificates
+RUN mkdir /usr/local/lib/python3.9/site-packages/rucio/daemons/hermes/kafka
```
Seems a bit strange to hard-code 3.9 here.  Is that because of how Rucio does things?

srp3rd replied:
Yes

ktlim commented on docker/rucio-daemons-hermesk/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+
+RUN mkdir -p /etc/grid-security/certificates
+RUN mkdir /usr/local/lib/python3.9/site-packages/rucio/daemons/hermes/kafka
+COPY ./lib/rucio/daemons/hermes/kafka /usr/local/lib/python3.9/site-packages/rucio/daemons/hermes/kafka/
```
It's better to copy the slowly-changing things first, then the rapidly-changing things.  That way the layer cache works better.

srp3rd replied:
Done

ktlim commented on etc/docker/dev/docker-compose-kafka.yml:
```diff
@@ -0,0 +1,440 @@
+version: "3"
```
Does this file get used anywhere?  If not, can it be removed?

srp3rd replied:
I use this file for internal testing

ktlim commented on lib/rucio/common/.github/workflows/build.yaml:
```diff
@@ -0,0 +1,43 @@
+name: Build Rucio Hermes Kafka container
```
This file isn't used either, is it?

srp3rd replied:
yes it is;  It is used to build the container that Brandon deploys.

ktlim commented on lib/rucio/common/schema/lsst.py:
```diff
@@ -0,0 +1,431 @@
+# -*- coding: utf-8 -*-
```
Is this just for testing?

srp3rd replied:
yes

ktlim commented on ups/ctrl_rucio_ingest.table:
```diff
@@ -0,0 +1,21 @@
+# List EUPS dependencies of this package here
```
Nothing here (or in the .cfg file) should be needed; I don't think you even want this to be an eups package.

srp3rd replied:
Removed

ktlim commented on lib/rucio/daemons/hermes/kafka/hermesk.py:
```diff
@@ -0,0 +1,766 @@
+# -*- coding: utf-8 -*-
```
I'm wondering if you can't express this as a patch on the source version that you have locked.  It would be a lot clearer than copying the entire file.  As it is, I can't easily tell what's your code and what's Rucio's.

srp3rd replied:
Understood.  After this PR is completed, I'll change this to a patch.

srp3rd replied:
Added a .txt patch file so you can see what was done.  The original hermes.py code was in release 33.0.0.  I tried to keep the edits to the same style as the original code.

https://github.com/lsst-dm/ctrl_rucio_ingest/blob/11a8060c77f87edce3a5e6c353b0c203edd51f1a/lib/rucio/daemons/hermes/kafka/hermesk_patch.txt

ktlim commented on lib/rucio/daemons/hermes/kafka/kafka_support.py:
```diff
@@ -0,0 +1,155 @@
+# -*- coding: utf-8 -*-
```
If we wrote this, it should have our copyright, not theirs.

srp3rd replied:
Fixed - in this file and the other original files.

srp3rd committed:
Update versions.env to 34.6.0
Update to new version 34.6.0 of constants.py
Update to new version of 34.6.0 protocol.py

srp3rd committed:
Add protocol.py

srp3rd committed:
Add hermesk patch info for mods to hermes.py in release 33.0.0

ktlim commented on lib/rucio/daemons/hermes/kafka/hermesk_patch.txt:
```diff
@@ -0,0 +1,115 @@
++                )
++                logger(
++                    logging.INFO,
++                    "%s messages processed by Kafka filter in %s seconds",
```
Might be good to log the number successful and failed.

srp3rd replied:
I'll mark that down for a new ticket, because it'll require some rework of the interface.

ktlim commented on lib/rucio/daemons/hermes/kafka/filters/filtered_list.py:
```diff
@@ -0,0 +1,99 @@
+
+class filtered_list(default):
+
+    RUBIN_BUTLER = "rubin_butler"
```
So this file is the really Rubin-specific component.  It wouldn't be part of the Rucio distribution, but we'd add it on for Rubin deployments, correct?

srp3rd replied:
yes

ktlim commented on lib/rucio/daemons/hermes/kafka/filters/filtered_list.py:
```diff
@@ -0,0 +1,99 @@
+            # and mark it for deletion.
+            if str(message['event_type']).lower() != 'transfer-done':
+                discard_count += 1
+                to_delete.append(message['id'])
```
I assume each configured messaging system gets to mark messages independently as "to be deleted" and that this is not preventing others from seeing the message.

srp3rd replied:
Yes, that's correct.  There are duplicate messages for each of the Rucio-configured systems they serve.  Each has an independent set of messages.

ktlim commented on lib/rucio/daemons/hermes/kafka/filters/filtered_list.py:
```diff
@@ -0,0 +1,99 @@
+                # get the metadata for this file
+                scope = str(message['payload'].get('scope'))
+                name = str(message['payload'].get('name'))
+                metadata = self.client.get_metadata(plugin='ALL', scope=scope, name=name)
```
There's no way to get only the Rubin additions rather than everything?

srp3rd replied:
We could, but sending everything lets the receiver decide what to do with the available contents of the message, rather than the sender.

ktlim replied:
But you aren't actually sending everything.  You're doing `metadata.get()` below to extract out the Rubin components we want.  I was just wondering if we could make this particular call more efficient by only retrieving the Rubin data.

srp3rd replied:
But it does send everything.   The retrieved message is augmented with the results of the `metadata.get()`  and the entire message is sent to Kafka.

ktlim commented on lib/rucio/daemons/hermes/kafka/filters/default.py:
```diff
@@ -0,0 +1,51 @@
+
+    def process(self, messages):
+        for topic in self.topics:
+            self._process(topic, messages)
```
You aren't accumulating and returning the list of processed messages.  Doesn't that mean that they won't be deleted and will instead be re-sent?

srp3rd replied:
Yes, this is a bug in default, which I'll correct. This is kind of a catch-22.   If you blindly delete messages that were successfully sent, you run the risk of deleting messages on topics that failed to send.  As the code stands, it should delete messages only if they're sent on all topics, and send duplicates out if messages failed to be sent on any of the topics.  This isn't a long term solution.  Possibilities are:  restrict to one topic for Kafka in default, or store messages in the DB specific to each topic.  This isn't an issue for filtered_list.py, because messages are sent on specific topics (and not all of them), and if they aren't sent, they aren't deleted.

ktlim replied:
The problem is that no messages are being deleted at all.  You're returning the list of `to_delete` from `_process()`, but then this code throws away that list rather than returning it as the result of `process()`.

I'm not exactly sure how this is interacting with the Kafka filter, but we are indeed getting hundreds of thousands of undeleted messages are piling up in the `messages` table.

srp3rd replied:
Yes, I understand the `to_delete` is an issue - what I was commenting on was that in addition to that problem, there's an additional problem of how to deal with messages sent to multiple topics, some of which can fail.  I'm going to create a new ticket to deal with that issue specifically.  I've added a temporary fix which addresses the `to_delete` list being thrown away.

The default filter doesn't interact with the Kafka filter in any way, and the hundreds of thousands of undeleted messages are either because HermesK wasn't running, or that additional services (like activemq) were configured, but not consumed properly by the base Rucio code.

srp3rd committed:
Remove ups directory and its contents

srp3rd committed:
Fix issue where messages weren't deleted in default filter. This needs
additional thought because if some topics aren't sent, messages could be
deleted before they topics become available. This might require additional
changes to Rucio itself to accomplish.

srp3rd committed:
Fix copyright notices

srp3rd committed:
Rework to use build-and-push-to-ghcr@v1

srp3rd committed:
Fix build.yaml to point at Dockerfile properly

srp3rd committed:
Fix syntax

srp3rd committed:
Fix for build file

srp3rd committed:
Add workflow_dispatch

srp3rd committed:
Move inputs to the right section

srp3rd committed:
Revert to old build

srp3rd committed:
Change ordering

srp3rd committed:
Force build
